gradle-baseline-java/src/main/groovy/com/palantir/baseline/plugins/BaselineImmutables.java

```diff
@@ -16,23 +16,33 @@
 
 package com.palantir.baseline.plugins;
 
+import java.util.Collections;
 import java.util.Objects;
 import org.gradle.api.Plugin;
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Dependency;
-import org.gradle.api.plugins.JavaPluginConvention;
 import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetContainer;
 import org.gradle.api.tasks.compile.JavaCompile;
 
 public final class BaselineImmutables implements Plugin<Project> {
 
     @Override
     public void apply(Project project) {
         project.getPluginManager().withPlugin("java", unused -> {
-            project.afterEvaluate(proj -> {
-                proj.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().stream()
-                        .filter(sourceSet -> hasImmutablesProcessor(project, sourceSet))
-                        .forEach(sourceSet -> addImmutablesIncrementalCompilerArg(project, sourceSet));
+            project.getExtensions().getByType(SourceSetContainer.class).configureEach(sourceSet -> {
```

Comment:
Can we use `.matching(...)` here to keep this fluent?

Reply:
Hmm, what would we match on here?

Reply:
`hasImmutablesProcessor`

Reply:
This is what we do in other plugins.

Reply:
I don't think this will be as lazy as it can be, as we currently check for immutables the moment the compiler args are requested, where as with `.matching` on the source sets would check for immutables the earliest of when each source set is created or when `configureEach` is called.

Also just be aware that there's a lot of really old stuff in this repo and a lot of it is a bit dodgy (but works well enough)

Reply:
I tend to avoid fluent code because it makes the behavior less obvious to readers, especially as ordering is concerned.

Reply:
Cool, make sense. Thanks for the explanation!